service/storage_service.cc

```diff
@@ -1003,6 +1005,70 @@ future<> storage_service::raft_bootstrap(raft::server& raft_server) {
     }
 }
 
+future<> storage_service::update_topology_with_local_metadata(raft::server& raft_server) {
+    // TODO: include more metadata here
+    auto local_shard_count = smp::count;
+    auto local_ignore_msb = _db.local().get_config().murmur3_partitioner_ignore_msb_bits();
+
+    auto synchronized = [&] () {
+        auto it = _topology_state_machine._topology.find(raft_server.id());
+        if (!it) {
+            throw std::runtime_error{"Removed from topology while performing metadata update"};
+        }
+
+        auto& replica_state = it->second;
+
+        return replica_state.shard_count == local_shard_count
+            && replica_state.ignore_msb == local_ignore_msb;
+    };
+
+    // We avoid performing a read barrier if we're sure that our metadata stored in topology
+    // is the same as local metadata. Note that only we can update our metadata, other nodes cannot.
+    //
+    // We use a persisted flag `must_update_topology` to avoid the following scenario:
+    // 1. the node restarts and its metadata changes
+    // 2. the node commits the new metadata to topology, but before the update is applied
+    //    to the local state machine, the node crashes
+    // 3. then the metadata changes back to old values and node restarts again
+    // 4. the local state machine tells us that we're in sync, which is wrong
+    // If the persisted flag is true, it tells us that we attempted a metadata change earlier,
+    // forcing us to perform a read barrier even when the local state machine tells us we're in sync.
```

Comment:
So @avikivity said that it is OK to requiring a quorum on boot, so we can do unconditional barrier and avoid the state all together.

Reply:
I'd prefer delaying that change - which is a pretty big one and impacts users - until we really deal with a problem which will be hard to solve without this requirement. I don't want to introduce it in this little PR. If we're sure about the decision at the moment of leaving experimental we can drop the state.

Reply:
BTW. we'll have a problem if we lose quorum permanently - even if we only require it in some situations
I opened an issue https://github.com/scylladb/scylladb/issues/13537